### DISCUSSION TITLE
Pre-commit hooks not running

### DISCUSSION THREAD

**Discussion by tnewhook** [gitextensions/gitextensions#10692] [Q&A]
### Environment

- Git Extensions 4.0.2.16100
- Build 25100ec1f7c8da8f798613da74826038af81a50e
- Git 2.39.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.13

### Issue description

Git extensions runs commit before pre-commit hook is done firing

### Steps to reproduce
-----------------------------------------------
create a pre-commit hook that calls a pre-commit.ps1 file 
#!/bin/sh
c:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe -ExecutionPolicy RemoteSigned -File '.git\hooks\pre-commit.ps1'

steps in in pre-commit powershell script:
1) check what's on the stage, 
2) If there's a *.zip file on stage, for each */zip file
  2a) extract zip file on stage
  2b) add extracted folder\* to stage
  2c) remove original .zip file from stage
3) exit 0, continuing with commit, which now has the extracted zip file without the original zip file(s)
------------------------------------------
create a zip file, run commit. 

Result: 
The commit doesn't wait for the hook to return 0, but commits the original zip file. I tried with command line git, which synchronously waits for the hook to return.

### Did this work in previous version of GitExtensions?

N/A

### Diagnostics

I didn't see anything related to calling the hook in the command log
[git command log.txt](https://github.com/gitextensions/gitextensions/files/10521524/git.command.log.txt)

**Comment by mstv** [comment]
Git calls the hooks, not GE.
GE has executed
```
git commit -F "c:/Repo/BI4-PBI/.git/COMMITMESSAGE"
```
Does this command really ignore the hook?

**Comment by tnewhook** [comment]
In the console window that pops up, I see the beginning of the script triggered, but the extracted files never show up in the folder.  The console window also exits really quickly, rather than waiting for the script to finish. Does git extensions do anything to kill the git process?

**Comment by mstv** [comment]
No, GE waits for git to terminate. Git *has* returned the exit code `0`. So it has not been killed by GE.
Perhaps, the powershell script is run detached. You could add a some sleep on start of the PS script.
Which shell have you used to run `git commit` manually? Compare with the `bash`, please.

**Comment by vbjay** [comment]
My guess is the unzipping is occuring by calling command that is not waiting for unzip to finish. So it is running and exiting and unzip is still running. 

Showing your hook can at least point is to validating the issue.

**Comment by tnewhook** [comment]
This is the entire hook: 

```
#!/bin/sh
c:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe -ExecutionPolicy RemoteSigned -File '.git\hooks\pre-commit.ps1'
```

I've used the Windows command line to do the git add, git commit, which runs the hook, and waits for the process to finish. 

The powershell script runs, then returns 0.

**Comment by mstv** [comment]
vbjay meant your powershell script, of course.

**Comment by vbjay** [comment]
Also an fyi. 

https://www.visualstudiogeeks.com/DevOps/UsingPowerShellForGitHooksWithVstsGitOnWindows

**Comment by vivainio** [comment]
Another problem is prepare-commit-msg hook. I have a hook that creates a well-formed commit message from branch name, but since it doesn't run I get empty commit message as starting point.

With e.g. "git gui", it runs and shows the message and allows you to edit it

**Comment by vbjay** [comment]
Let me show you the right way.  As you can see the hook bash script just says it is a powershell hook and git knows to run the same named ps1 file.  I did point you to this in the https://github.com/gitextensions/gitextensions/discussions/10692#discussioncomment-4835707 comment.
[pre-commit.zip](https://github.com/gitextensions/gitextensions/files/13439650/pre-commit.zip)

![image](https://github.com/gitextensions/gitextensions/assets/3629489/9e46496f-25a5-40e5-ac18-9b70a0acae0f)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/0a77897c-119b-4b93-b008-2bdd5654122a)
So Like I asked before.  I needed to see the actual hook code and not the go run this script.  As you can see the hook ran just fine.  Your issue is you are running the unzipping process as a new process and in the hook  you aren't waiting for the process to end.
